### PR TITLE
feat(machines): server-authoritative workspace list + layout sync

### LINKED ISSUE #2048: Server-synced multiplayer workspaces (machine workspace list/layout/active-selection)
## Context

PR #2031 wires the shared `MachineTree` sidebar to `useMachineWorkspaceStore`, giving each machine a persistent set of named workspaces (sidebar items owning a pane grid). It includes cross-browser **session** discovery: a running agent-terminal session spawned from another browser/user shows up as an "unclaimed" row (diffed against `/api/machines/agent-terminals`) and can be adopted into a local workspace.

## Gap

Discovery stops at the session layer. The workspace itself — names, grid/pane layout, and which workspace is "active" — lives entirely in a Zustand store persisted to `localStorage` only (`apps/web/src/stores/machine-workspace/useMachineWorkspaceStore.ts`, `name: 'machine-workspace-storage'`, no custom `storage` adapter). It never touches the server or `apps/realtime`'s Socket.IO layer.

Concretely, for two users (or two browser tabs of the same user) pointed at the same machine:
- Workspace lists are independent — different names, different counts, no shared source of truth.
- No live push when someone else creates, renames, or closes a workspace.
- "Active workspace" selection is per-browser, not shared — there's no way to see what a teammate is currently looking at.
- The only thing that reconciles across browsers is the *unclaimed session* row, and only on mount/expand (poll-based via SWR, not pushed).

For a product positioned as multiplayer, this means "multiplayer" currently only exists at the terminal-session layer, not the workspace-organization layer.

## Proposed follow-up

- Persist workspace metadata (names, grid layout, active-workspace-per-machine) server-side instead of (or in addition to) `localStorage`.
- Broadcast workspace create/rename/delete/activate events over the existing `apps/realtime` Socket.IO channel so open browsers update live.
- Decide on conflict/ownership semantics: is "active workspace" per-user or a single shared cursor per machine? (Likely per-user, but the workspace *list* itself should be shared.)
- Reconcile with the existing `unclaimed session` adoption flow so it becomes redundant once workspaces are the shared source of truth (a session should just show up under its real workspace, not as an orphan row).

## Non-goals for this issue

- Real-time terminal I/O sharing (already works via the existing PTY/socket plumbing) — this is scoped to workspace *organization* metadata only.

---
Filed from PR #2031 review discussion.

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

Closes #2048. Makes a Machine's workspace list (create/rename/delete) and each
workspace's pane-grid layout **server-authoritative**, broadcast live over the
existing `apps/realtime` Socket.IO room for the Machine's own page id — closing
the gap left by PR #2031, where workspace metadata lived only in that browser's
`localStorage`.

- **Schema**: `machine_workspaces` (client-minted `id`, explicit `scope`
  discriminant, `layout` jsonb holding only the *shared* structural half of the
  grid) + `machine_workspace_bootstraps` (a one-row-per-machine atomic claim so
  exactly one browser's local history ever seeds the server on first load —
  per-id upsert alone doesn't prevent N racing browsers' disjoint lists from
  all landing and being unioned together).
- **Service + API**: `packages/lib/services/machines/machine-workspaces{,-store}.ts`
  + `GET/POST/PATCH/DELETE /api/machines/workspaces` + `POST .../bootstrap`,
  following the existing `machine-projects` pattern. Create is an idempotent
  upsert-by-id (first-writer-wins — two browsers materializing the same
  session-derived workspace is expected, not an error).
- **Realtime**: `machine-workspace:{created,updated,deleted,bootstrapped}`
  events via the existing signed `/api/broadcast` HTTP endpoint (no new
  room-join logic needed — a Machine's identity is its own page id).
- **Client**: `useMachineWorkspaceSync` (mounted once at `MachineView`, the
  true per-machine root) hydrates from the server, runs the bootstrap-once
  flow, and reconciles live broadcasts into the existing Zustand store.
  `useSyncedWorkspaceActions` wraps the store's identity/layout actions
  (create/rename/split/close/bindPaneTerminal/openTerminal) to push the
  resulting state to the server — PATCH first, falling back to POST-create on
  a 404, so the wrapper needs no state shared across the multiple components
  that call it (sidebar + pane grid).
- `activeWorkspaceId`, `activePaneId`, `pendingPickerPaneId`, and a pane's
  `pendingPrompt` all stay **local per browser**, never synced — presence-like
  ("what am I looking at / focused on"), explicitly out of scope per the issue.
- Minimal rename UI: double-click a workspace's name in `WorkspaceLeaves` to
  edit it inline (no rename affordance existed before this PR).
- `sessionWorkspaceId`'s delimiter changed from a NUL character to the Unit
  Separator control character (U+001F) — the id is now also a Postgres
  primary key, and Postgres `text` rejects a literal NUL byte outright.
- The existing "unclaimed session" reconciliation in `WorkspaceLeaves` is kept
  unchanged (not made redundant by this PR — it bridges a different gap: a
  `machine_agent_terminals` row with no owning workspace, which can still
  happen via a spawn path that bypasses `openTerminal` entirely).

## Test plan

- [x] Schema-level tests (no DB) for both new tables
- [x] Service-level tests: idempotent create-upsert race, update/rename,
      remove, and the bootstrap claim race (first caller wins, second gets the
      winner's list back unseeded)
- [x] Route contract tests for all 5 endpoints (auth gating, denial-to-status
      mapping, broadcast-only-on-real-change)
- [x] Reducer/store tests: `sessionWorkspaceId`'s new delimiter,
      `mergeServerWorkspaces`/`applyServerWorkspaceUpsert`/`applyServerWorkspaceDeleted`
      preserve local-only fields and never keep unpublished local stragglers
- [x] Hook tests for `useMachineWorkspaceSync` (hydrate vs. bootstrap
      branching, event filtering by machineId) and `useSyncedWorkspaceActions`
      (PATCH-then-POST-fallback)
- [x] `bun run typecheck` and `bun run lint` clean across the monorepo
- [x] Full `packages/db`, `packages/lib`, `apps/web` vitest suites green
      (13101+ passing); the only failures are pre-existing, unrelated to this
      change (a missing local Postgres role for a couple of DB-integration
      tests, and one pre-existing flaky timezone-dependent test) — confirmed
      via `git status` that none of the failing files were touched here

## Review convergence

Fixes made in response to Codex + CodeRabbit review passes, all with regression
test coverage:

- **Cross-machine id collision (P1)**: `machine_workspaces`' primary key was
  the lone `id` column, but `sessionWorkspaceId` has no `machineId` in it — two
  different Machines reusing the same project/branch/session name could
  collide. Now a compound `(machineId, id)` primary key.
- **Legacy NUL-delimited id migration (P1)**: existing `localStorage` from
  PR #2031 has NUL-delimited `sessionWorkspaceId`s; Postgres `text` rejects
  literal NUL bytes. `migrateLegacyWorkspaceId` rewrites them to the current
  delimiter before they ever reach the bootstrap POST.
- **Bootstrap claim cascade (Critical)**: `machine_workspace_bootstraps`'
  `bootstrappedByUserId` FK was `notNull` + `cascade` — deleting the winning
  user's account would delete the claim row and reopen the exact
  duplicate-bootstrap race the table exists to prevent. Now nullable +
  `set null`.
- **Partial `updated` events**: an `updated` socket event unconditionally
  upserted into the store even for a workspace this browser hadn't hydrated
  yet, defaulting missing fields (including `scope`, which the event never
  carries) — could plant a broken zero-column, wrong-scope entry. Now
  unconditionally skipped when the workspace is unknown locally.
- **Rename `onBlur` race**: in Chromium, unmounting the focused rename input
  (which Escape/Enter both do) can itself fire a native `blur`, resurrecting a
  cancelled draft or double-committing. Fixed with a `skipNextBlur` guard.
- **`useMachineWorkspaceSync` gating**: was running (fetch + socket join) for
  non-admin viewers who never see the tabs it exists for. Now takes
  `machineId: string | null`; `MachineView` passes `null` for non-admins.
- **PATCH-fallback detection**: was string-matching `error.message ===
  'not_found'`, coupled to `fetchJSON`'s exact error construction. Now checks
  the real `response.status === 404` via `fetchWithAuth` directly.
- **Partial PATCH per field**: every synced action sent both `name` and
  `columns` on every PATCH, even though the route supports updating either
  independently — a rename racing a split (each on its own possibly-stale
  copy of the field it didn't touch) could silently revert the other's
  change. Each action now sends only the field it actually changed.
- **Batched last-workspace pane cleanup**: emptying every running pane when
  removing a machine's only workspace fired one independent PATCH per pane
  with no ordering guarantee. New `closePanes` applies every local close
  first, then pushes exactly one PATCH with the final layout.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Br1TJjStaT9y1CkURf2uTj

## Summary by CodeRabbit

* **New Features**
  * Added server-synchronized machine workspaces with create, rename, layout updates, deletion, and one-time bootstrapping.
  * Enabled real-time workspace synchronization across machine views and the development sidebar.
  * Added inline workspace renaming with Enter, blur, and Escape controls.
  * Introduced durable workspace storage with automatic migration of legacy workspace identifiers.
* **Bug Fixes**
  * Preserved local pane/UI state during server synchronization and reconciliation.
  * Prevented duplicate updates/broadcasts and fixed rename interaction edge cases.
* **Tests**
  * Expanded contract, integration, and UI regression coverage for workspaces APIs and inline editing.

### CODE REVIEW

File: packages/lib/src/services/machines/machine-workspaces-store.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+          name: input.name,
+          layout: input.layout,
+          createdAt: input.now,
+          updatedAt: input.now,
+        })
+        .onConflictDoNothing({ target: machineWorkspaces.id })
+        .returning();
+      if (inserted) return { created: true, row: toRecord(inserted) };
+
+      const [existing] = await db.select().from(machineWorkspaces).where(eq(machineWorkspaces.id, input.id)).limit(1);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/services/machines/machine-workspaces-store.ts]: Confirmed and fixed in 015772534.

The primary key was the lone `id` column, but `sessionWorkspaceId` derives purely from `project/branch/session-name` text with no `machineId` in it — two different Machines reusing the same names would collide. Fixed by making the primary key the COMPOUND `(machineId, id)`:

- Schema: `primaryKey({ columns: [table.machineId, table.id] })` instead of `.primaryKey()` on `id` alone.
- Both `onConflictDoNothing` targets (`insertIfAbsent` and `bootstrapSeed`'s per-row insert) now target `[machineWorkspaces.machineId, machineWorkspaces.id]`.
- The fallback SELECT in `insertIfAbsent` (the line you flagged) is now scoped by `and(eq(machineId), eq(id))`, not `id` alone.
- Regenerated migration 0205 in place (it was never applied anywhere yet, so no ALTER was needed).
- Added a regression test proving two different machines minting the identical id both succeed independently as two distinct rows.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/stores/machine-workspace/workspace-reducer.ts [hunk 2]
```diff
@@ -545,11 +673,15 @@ export function sanitizeMachines(value: unknown): Record<string, MachineWorkspac
...
+ * workspace, with whatever panes were split into it still there. Joined with
+ * U+001F (Unit Separator): project and branch names can contain '/' and ':',
+ * so an ordinary character can't be used as the delimiter. NOT NUL (U+0000)
+ * — this id is also the primary key of the server-side `machine_workspaces`
+ * row (see `machine-workspaces-runtime.ts`), and Postgres `text` columns
+ * reject a literal NUL byte outright.
  */
 export function sessionWorkspaceId(scope: OpenTerminalScope): string {
-  return `session\u0000${scope.projectName ?? ''}\u0000${scope.branchName ?? ''}\u0000${scope.name}`;
+  return `session\u001f${scope.projectName ?? ''}\u001f${scope.branchName ?? ''}\u001f${scope.name}`;
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/stores/machine-workspace/workspace-reducer.ts:714]: Confirmed and fixed in 015772534.

Added `migrateLegacyWorkspaceId` inside `sanitizeMachines` (the existing "what comes back from storage is untrusted" scrubber) — any workspace id, `order` entry, or `activeWorkspaceId` still using the old NUL delimiter gets rewritten to the current U+001F one, on load, before it ever reaches `useMachineWorkspaceSync`'s bootstrap POST. It's a plain 1:1 character substitution, so it preserves the "same session, same id" property and is a no-op for `crypto.randomUUID()`-based ids (which never contain either character).

Added a regression test in `workspace-reducer.test.ts` covering the exact scenario: a persisted workspace whose id, record key, `order`, and `activeWorkspaceId` all still use the legacy delimiter get migrated consistently everywhere.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/components/layout/middle-content/page-views/machine/MachineView.tsx [hunk 3]
```diff
@@ -52,12 +53,19 @@ const TAB_TRIGGERS: { value: MachineTabValue; label: string; icon: React.Element
...
+ * (Radix `TabsContent` unmounts inactive tabs; this component doesn't), so the
+ * socket room join / server hydration / bootstrap race only happen once per
+ * machine, not once per Terminal-tab activation (#2048).
  */
 const MachineView = ({ pageId, embedded = false }: MachineViewProps) => {
   const { user } = useAuth();
   const isAdmin = user?.role === 'admin';
   const activeTab = useMachineTabStore((state) => state.tabs[pageId] ?? DEFAULT_MACHINE_TAB);
   const setTab = useMachineTabStore((state) => state.setTab);
+  useMachineWorkspaceSync(pageId);
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/machine/MachineView.tsx:72]: Confirmed and fixed in 9dee53153.

`useMachineWorkspaceSync` now takes `machineId: string | null` — `MachineView` passes `isAdmin ? pageId : null` instead of always calling it with `pageId`. Every internal effect/SWR-key/socket-subscription now guards on `machineId` being non-null, so a non-admin viewer triggers no workspace fetch and no socket room join.

Called with `null` rather than skipping the hook call — React's rules of hooks forbid a conditional hook call, and this hook is unconditionally mounted at `MachineView`, the one true per-machine root.

Added regression coverage in `MachineView.test.tsx`: one test asserting a non-admin passes `null`, one asserting an admin passes the real `pageId`.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/components/layout/middle-content/page-views/machine/workspace/WorkspaceLeaves.tsx [hunk 4]
```diff
@@ -116,10 +119,21 @@ export default function WorkspaceLeaves({
...
+    setRenamingId(workspace.id);
+    setDraftName(workspace.name);
+  };
+
+  const commitRename = (workspaceId: string, currentName: string) => {
+    const trimmed = draftName.trim();
+    if (trimmed && trimmed !== currentName) renameWorkspace(workspaceId, trimmed);
+    setRenamingId(null);
+  };
+
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/machine/workspace/WorkspaceLeaves.tsx:178]: Confirmed and fixed in 9dee53153.

Added a `skipNextBlur` ref in `WorkspaceLeaves.tsx`. Escape (`cancelRename`) and Enter (`commitRenameViaKey`) both set it to `true` right before closing the editor; the next `onBlur` (`commitRenameViaBlur`) checks it, consumes it, and returns early instead of re-entering `commitRename`. It's reset to `false` when a fresh rename starts, so a stale `true` can never suppress a later legitimate blur-to-commit.

Added two regression tests using `fireEvent.keyDown` + `fireEvent.blur` to simulate the Chromium unmount-fires-blur quirk directly: one proving a blur right after Escape doesn't resurrect the cancelled draft, one proving a blur right after Enter doesn't fire a second PATCH.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/hooks/useMachineWorkspaceSync.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      // `updated` only carries whichever of name/columns actually changed —
+      // fill in the rest from what this browser already has for it.
+      const current = useMachineWorkspaceStore.getState().machines[machineId]?.workspaces[payload.workspaceId];
+      applyServerUpsert(machineId, {
+        id: payload.workspaceId,
+        name: payload.name ?? current?.name ?? '',
+        scope: current?.scope ?? {},
+        columns: payload.columns ?? current?.columns ?? [],
+      });
+    };
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/hooks/useMachineWorkspaceSync.ts:190]: Confirmed and fixed in 9dee53153 — with a broader guard than the suggested diff.

The suggested fix still upserts when `current` is undefined but the payload happens to include BOTH `name` and `columns`. But `updated` events never carry `scope` at all (it's immutable, so the PATCH route never re-sends it) — so even a 'fully described' partial payload has no safe `scope` to fall back on for a workspace this browser doesn't know yet. Simplified to: `if (!current) return;` unconditionally — any update for an unknown workspace is skipped, full stop, and a later `created` broadcast or hydration introduces it correctly instead.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/hooks/useMachineWorkspaceSync.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+  try {
+    await patch('/api/machines/workspaces', { machineId, workspaceId, name: workspace.name, columns });
+  } catch (error) {
+    // Not found server-side yet — create it. `updateWorkspace`'s route returns
+    // `{error: 'not_found', reason: 'not_found'}` on a 404, which `auth-fetch.ts`
+    // surfaces as an Error whose message IS that reason string. Any OTHER
+    // failure (network blip, 500) is swallowed here: the local grid already
+    // reflects the user's action, and the next successful push (this browser's
+    // own, or a broadcast from elsewhere) reconciles the server.
+    if (!(error instanceof Error) || error.message !== 'not_found') return;
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/hooks/useMachineWorkspaceSync.ts]: Confirmed and fixed in 9dee53153.

`pushWorkspaceUpdate` now calls `fetchWithAuth` directly instead of the `patch()" helper, so it gets the real `Response` object and branches on `response.status === 404` — no coupling to `fetchJSON`'s error-message construction or the 404 body shape at all.

Updated tests to mock `fetchWithAuth` returning `{ok: false, status: 404}` / `{status: 500}` instead of a rejected `patch()` promise, plus a new test confirming no POST fallback when the PATCH succeeds.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: packages/db/src/schema/machine-workspace-bootstraps.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  machineId: text('machineId')
+    .primaryKey()
+    .references(() => pages.id, { onDelete: 'cascade' }),
+
+  bootstrappedByUserId: text('bootstrappedByUserId')
+    .notNull()
+    .references(() => users.id, { onDelete: 'cascade' }),
+
+  bootstrappedAt: timestamp('bootstrappedAt', { mode: 'date' }).defaultNow().notNull(),
+});
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/db/src/schema/machine-workspace-bootstraps.ts:41]: Confirmed and fixed in 9dee53153 — same fix as your suggested diff.

`bootstrappedByUserId` is now nullable with `onDelete: 'set null'` instead of `notNull` + `cascade`. Added a doc comment explaining why: this row's mere EXISTENCE is the load-bearing invariant (it closes the duplicate-bootstrap race), so the winning user's account deletion must not un-claim the machine.

Updated the schema test: asserts `notNull === false` on that column, and asserts the two FKs' `onDelete` independently by column (`machineId` → cascade, `bootstrappedByUserId` → set null) rather than asserting both cascade.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/hooks/useMachineWorkspaceSync.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+ * after creation): PATCH first, falling back to POST-create on a 404 — a brand
+ * new workspace materialized via `openTerminal`'s "existing session, new pane"
+ * branch has never gone through `createWorkspace`'s own POST. */
+async function pushWorkspaceUpdate(machineId: string, workspaceId: string): Promise<void> {
+  const workspace = useMachineWorkspaceStore.getState().machines[machineId]?.workspaces[workspaceId];
+  if (!workspace) return;
+  const columns = toWireColumns(workspace.columns);
+
+  try {
+    await patch('/api/machines/workspaces', { machineId, workspaceId, name: workspace.name, columns });
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/hooks/useMachineWorkspaceSync.ts]: Confirmed and fixed in 4e614f74b.

The route (and `updateWorkspace`) already supported a true partial update, but `pushWorkspaceUpdate` always sent both `name` and `columns` regardless of which one the calling action actually changed — exactly the clobber scenario you described. Each synced action now passes which field(s) it touched (`{name: true}` for rename, `{columns: true}` for every layout-affecting action), and `pushWorkspaceUpdate` only includes those in the PATCH body.

Added two regression tests asserting the PATCH body for `splitRight` has `columns` but not `name`, and vice versa for `renameWorkspace`.

Left open for your/a human reviewer's verification rather than resolving it myself.

File: apps/web/src/components/layout/middle-content/page-views/machine/workspace/WorkspaceLeaves.tsx [hunk 9]
```diff
@@ -180,9 +218,9 @@ export default function WorkspaceLeaves({
           // instead of dropping the (unchanged) workspace, or it would linger
           // bound to the agent names just killed above.
           if (workspacesOf(machine).length <= 1) {
-            pendingRunningPaneIds.forEach((paneId) => closePane(machineId, pendingRemove, paneId));
+            pendingRunningPaneIds.forEach((paneId) => closePane(pendingRemove, paneId));
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/components/layout/middle-content/page-views/machine/workspace/WorkspaceLeaves.tsx]: Confirmed and fixed in 4e614f74b.

Added `closePanes(workspaceId, paneIds)` to `useSyncedWorkspaceActions`: it applies every local `closePane` first (synchronous), then fires exactly ONE PATCH with the resulting final layout, instead of one independent fire-and-forget PATCH per pane. `WorkspaceLeaves.tsx`'s last-workspace cleanup path now calls this instead of `.forEach(closePane)`.

Added a regression test: splits a workspace into 2 panes, closes both via `closePanes`, and asserts exactly one PATCH call fires with a body reflecting the final (fully-collapsed) layout.

Left open for your/a human reviewer's verification rather than resolving it myself.
